File: dSoak-Common-CSharp/SharedObjects/PublicEndPoint.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Runtime.Serialization;
using System.Net;

namespace SharedObjects
{
    [DataContract]
    public class PublicEndPoint
    {
        private IPAddress myIPAddress = null;
        private IPEndPoint myIPEndPoint = null;
        private string myHost = null;
        private Int32 myPort = 0;

        [DataMember]
        public string HostAndPort
        {
            get { return (myHost == null) ? "0.0.0.0:" + myPort.ToString() : myHost.ToString() + ":" + myPort.ToString(); }
            set { SetHostAndPort(value); }
        }

        public string Host
        {
            get { return myHost; }
            set { myHost = value; myIPAddress = null; myIPEndPoint = null; }
        }

        public Int32 Port
        {
            get { return myPort; }
            set { myPort = value; myIPEndPoint = null; }
        }

        public PublicEndPoint() { }

        public PublicEndPoint(string hostnameAndPort)
        {
            if (!string.IsNullOrWhiteSpace(hostnameAndPort))
            {
                string[] tmp = hostnameAndPort.Split(':');
                if (tmp.Length == 2 && !string.IsNullOrWhiteSpace(tmp[0]))
                {
                    Host = tmp[0];
                    Int32 tmpPort = 0;
                    Int32.TryParse(tmp[1], out tmpPort);
                    Port = tmpPort;
                }
            }
        }

        public IPEndPoint IPEndPoint
        {
            get
            {
                if (myIPEndPoint == null)
                {
                    if (myIPAddress == null)
                    {
                        if (!string.IsNullOrWhiteSpace(Host))
                            myIPAddress = LookupAddress(Host);
                    }
                    if (myIPAddress != null)
                        myIPEndPoint = new IPEndPoint(myIPAddress, Port);
                }
                return myIPEndPoint;
            }
            set
            {
                if (value != null)
                {
                    Host = value.Address.ToString();
                    Port = value.Port;
                    myIPAddress = value.Address;
                    myIPEndPoint = value;
                }
            }
        }

        public static IPAddress LookupAddress(string host)
        {
            IPAddress result = null;
            if (!string.IsNullOrWhiteSpace(host))
            {
                IPAddress[] addressList = Dns.GetHostAddresses(host);
                for (int i = 0; i < addressList.Length && result == null; i++)
                    if (addressList[i].AddressFamily == System.Net.Sockets.AddressFamily.InterNetwork)
                        result = addressList[i];
            }
            return result;
        }

        private void SetHostAndPort(string hostnameAndPort)
        {
            if (!string.IsNullOrWhiteSpace(hostnameAndPort))
            {
                string[] tmp = hostnameAndPort.Split(':');
                if (tmp.Length == 2 && !string.IsNullOrWhiteSpace(tmp[0]))
                {
                    Host = tmp[0];
                    Int32 tmpPort = 0;
                    Int32.TryParse(tmp[1], out tmpPort);
                    Port = tmpPort;
                }
            }
        }

        public override string ToString()
        {
            return string.Format("{0}:{1}", Host, Port);
        }
    }
}

File: dSoak-Common-Java/src/SharedObject/PublicEndPoint.java
package SharedObject;

import java.io.Serializable;
import java.net.InetAddress;
import java.net.UnknownHostException;

public class PublicEndPoint implements Serializable
{
	private static final long serialVersionUID = 1L;
	private InetAddress myIPAddress= null;
	private IPEndPoint myIPEndPoint = null;
	private String myHost;
	private int myPort;

	public PublicEndPoint() {	}
	
	public PublicEndPoint(String hostnameAndPort) 
	{
		HostAndPort(hostnameAndPort);
	}
	
	public String HostAndPort()
	{
		return ((myHost == null) ? ("0.0.0.0:" + myPort) : myHost.toString() + ":" + myPort);
	}
	
	public void HostAndPort(String value)
	{
		SetHostAndPortValue(value);
	}
	
	public String Host()
	{
		return myHost;
	}
	
	public void Host(String value)
	{
		myHost = value;
		myIPAddress = null;
		myIPEndPoint = null;
	}
	
	public int Port()
	{
		return myPort;
	}
	
	public void Port(int value)
	{
		myPort = value;
		myIPEndPoint = null;
	}
	
	private void SetHostAndPortValue(String str) 
	{
		if (str != null)
		{
			String[] tmp = str.split(":");
			if ((tmp.length == 2) && (tmp[0] != null))
			{
				Host(tmp[0]);
				Port(Integer.parseInt(tmp[1]));
			}
		}
	}

	public IPEndPoint IPEndPoint() throws UnknownHostException
    {
		if (myIPEndPoint == null) 
		{
			if  (myIPAddress == null)
			{
				if (Host() != null)
				{
					myIPAddress = LookupAddress(Host());
				}
				if  (myIPAddress != null)
				{
					myIPEndPoint =  new IPEndPoint(myIPAddress, Port());
				}
			}
		}
		return myIPEndPoint;
    }
	
	public void IPEndPoint(IPEndPoint value)
    {
		if (value != null)
        {
			Host(value.Address.getHostAddress());
            Port(value.Port);
            myIPAddress = value.Address;
            myIPEndPoint = value;
         }
    }
	
	public static InetAddress LookupAddress(String host) throws UnknownHostException
    {
		InetAddress result = null;
		InetAddress[] addressList = InetAddress.getAllByName(host);
		
        for (int i = 0; i < addressList.length && result == null; i++)
        {
        	String str1 = addressList[i].getHostName();
        	String str2 = addressList[i].getHostAddress();
        	if (str1.equals(host))
                result = InetAddress.getByName(str2);
        }
        return result;
    }
	
	@Override
	public String toString()
	{
		return Host() + ":" + Port();
	}
}